### PR TITLE
RUN-985: Add WebAssembly SIMD documentation

### PR DESCRIPTION
**Pull request by dfinity-berestovskyy**
Closes RUN-985

### CODE REVIEW

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+There are two main ways to benefit from WebAssembly SIMD in a smart contract:
+
+1. **Loop auto-vectorization:** Just enabling the WebAssembly SIMD and recompiling the project might be enough to get a significant performance boost. This is usually simple, error-proof, and often a one-line change. This is often the recommended first step, but the result depends heavily on the used algorithms, libraries, and compilers.
+
+2. **SIMD intrinsics:** Some computation-heavy functions might be rewritten using direct SIMD instructions. This opens the full SIMD potential, but in many cases some core canister algorithms must be completely rewritten using new instructions.
+
+## Using Loop Auto-Vectorization
+
+To leverage the loop auto-vectorization, the WebAssembly SIMD instructions should be enabled globally for the entire workspace or locally for specific functions within the canister. Once the instructions are available to the compiler, it automatically converts some normal loops into loops with parallel computations.
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:52]:
```suggestion
To leverage the loop auto-vectorization, the WebAssembly SIMD instructions should be enabled globally for the entire workspace, or locally for specific functions within the canister. Once the instructions are available to the compiler, it automatically converts some normal loops into loops with parallel computations.
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+While the change is easy and error-proof, the result in practice depends on many factors, like the algorithm itself, the compiler optimization level and options, project dependencies, etc.
+
+### Enabling SIMD Example
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:56]:
```suggestion
### Example
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+To enable WebAssembly SIMD instructions globally, for the whole workspace and all its dependencies:
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:58]:
```suggestion
To enable WebAssembly SIMD instructions globally for the whole workspace and all its dependencies:
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+<AdornedTabs groupId="languages">
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:60]: Using tabs is not necessary if there is only one language being shown - are there any other language examples that will be added?

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+#[target_feature(enable = "simd128")]
+fn auto_vectorization() {
+    ...
+}
+```
+
+</TabItem>
+</AdornedTabs>
+
+Note, WebAssembly SIMD instructions might be enabled by default in future `dfx` versions, so enabling it for a specific function within a canister might have no effect.
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:91]:
```suggestion
:::info
WebAssembly SIMD instructions may be enabled by default in future `dfx` versions, so enabling it for a specific function within a canister might have no effect.
:::
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+## Using WebAssembly SIMD Intrinsics
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:93]:
```suggestion
## Using WebAssembly SIMD intrinsics
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+WebAssembly SIMD instructions are available as platform-specific intrinsics for the `wasm32` platform. To use the intrinsics, the WebAssembly SIMD instructions should be enabled as described in the previous section.
+
+### Using SIMD Intrinsics Example
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:97]:
```suggestion
### Example
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    let mut res = [0.0; 4];
+    unsafe { v128_store(res.as_mut_ptr() as *mut v128, c) };
+    res
+}
+```
+
+</TabItem>
+</AdornedTabs>
+
+## Frequently Asked Questions
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:127]:
```suggestion
## Frequently asked questions
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+### How to measure performance speedup of a canister?
+
+The Internet Computer Protocol provides the `ic0.performance_counter` [System API call](https://internetcomputer.org/docs/current/references/ic-interface-spec#system-api-performance-counter) to measure a canister's performance.
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:131]:
```suggestion
ICP provides the `ic0.performance_counter` [system API call](https://internetcomputer.org/docs/current/references/ic-interface-spec#system-api-performance-counter) to measure a canister's performance.
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+There is also the `canbench` [benchmarking framework](https://github.com/dfinity/canbench) for canisters on the Internet Computer.
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:133]:
```suggestion
There is also the `canbench` [benchmarking framework](https://github.com/dfinity/canbench).
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+### Are there any libraries for Artificial Intelligence (AI) inference?
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:135]:
```suggestion
### Are there any libraries for artificial intelligence (AI) inferences?
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+Sonos `tract` is a tiny, self-contained, Tensorflow and ONNX inference [Rust library](https://github.com/sonos/tract). DFINITY [contributed WebAssembly SIMD support](https://github.com/sonos/tract/pull/1420) to the library. The library is used in some DFINITY AI [demos](https://www.youtube.com/watch?v=6qLvIXiCGcM) and [examples](https://github.com/dfinity/examples/tree/master/rust/image-classification).
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:137]:
```suggestion
The Sonos `tract` is a tiny, self-contained, Tensorflow and ONNX inference [Rust library](https://github.com/sonos/tract). DFINITY [contributed WebAssembly SIMD support](https://github.com/sonos/tract/pull/1420) to the library. The library is used in some DFINITY AI [demos](https://www.youtube.com/watch?v=6qLvIXiCGcM) and [examples](https://github.com/dfinity/examples/tree/master/rust/image-classification).
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+## References and Examples
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:139]:
```suggestion
## References and examples
```

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+* [WebAssembly SIMD Rust example](https://github.com/dfinity/examples/tree/master/rust/simd) compares the performance of a naive, optimized, auto-vectorized and SIMD intrinsic matrix multiplication running on the ICP blockchain.
```
Comment (jessiemongeon1) [inline docs/developer-docs/smart-contracts/advanced-features/simd.mdx:141]:
```suggestion
* [WebAssembly SIMD Rust example](https://github.com/dfinity/examples/tree/master/rust/simd) compares the performance of a naive, optimized, auto-vectorized and SIMD intrinsic matrix multiplication running on ICP.
```

**Review by jessiemongeon1 — APPROVED** [review]

File: docs/developer-docs/smart-contracts/advanced-features/simd.mdx (hunk 4 above)
Comment (dfinity-berestovskyy) [reply docs/developer-docs/smart-contracts/advanced-features/simd.mdx:60]: AFAIK, no SIMD plans for Motoko. Maybe there will be something from JS/Python side...

Comment (dfinity-berestovskyy) [reply docs/developer-docs/smart-contracts/advanced-features/simd.mdx:58]: I'll try to merge now as is and see later if there will be more examples...

Comment (jessiemongeon1) [reply docs/developer-docs/smart-contracts/advanced-features/simd.mdx:58]: Sounds good, thanks!

**Pull request merged**